Add-ons/UmbracoForms/Developer/Extending/Adding-a-Prevaluesourcetype.md

```diff
@@ -0,0 +1,70 @@
+---
+versionFrom: 9.0.0
+versionTo: 10.0.0
+---
+
+# Adding a Prevalue Source Type to Umbraco Forms
+
+The following example shows an illustrative custom prevalue source type that returns a hard-coded list of values. It can be extended for your needs via injection of services via the constructor. Settings can be applied and validated as shown in the other example custom types in this section.
+
+```csharp
+using System;
+using System.Collections.Generic;
+using Umbraco.Forms.Core;
+using Umbraco.Forms.Core.Models;
+
+namespace MyFormsExtensions
+{
+    public class FixedListPrevalueSource : FieldPreValueSourceType
+    {
+        public FixedListPrevalueSource()
+        {
+            Id = new Guid("42C8158D-2AA8-4621-B653-6A63C7545768");
+            Name = "Fixed List";
+            Description = "Example prevalue source providing a fixed list of values.";
+        }
+
+        public override List<PreValue> GetPreValues(Field field, Form form) =>
```

Comment:
@AndyButland should the parameters be nullable as they are in Umbraco 10 and I get some warning when it override the method where `Field` and `Form` isn't nullable.

![image](https://user-images.githubusercontent.com/2919859/178259894-e2044bd6-5426-41ba-8370-25d8ecea14e0.png)

Reply:
That's right yes.  I did this example for a V9 project.